backend/accounts/models.py

```diff
@@ -0,0 +1,10 @@
+from django.contrib.auth.models import AbstractUser
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+
+class CustomUser(AbstractUser):
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = []
```

Comment:
`REQUIRED_FIELDS` je [defaultne rovno](https://github.com/django/django/blob/master/django/contrib/auth/models.py#L362) `['email']`, ty to "nulujes". Procpak to? (policko email povinne byt musi, protoze z nej delame username predchozim zapisem; nebo mi neco unika?)

Reply:
```
ERRORS:
accounts.CustomUser: (auth.E002) The field named as the 'USERNAME_FIELD' for a custom user model must not be included in 'REQUIRED_FIELDS'.
	HINT: The 'USERNAME_FIELD' is currently set to 'email', you should remove 'email' from the 'REQUIRED_FIELDS'.
```

Hází mi to tuto chybu, také jsem úplně nerozměl té logice, proč nemůže být USERNAME_FIELD v `required_fields`. Snažil jsem se nějak googlit proč, ale nic moc jsem nedohledal viz: https://code.djangoproject.com/ticket/20097

Koukal jsem kde všude se to v djangu používá a používá se to je to pouze na dvou místech:

`django.contrib.auth.checks.py` --> Pouze kontroluje UserModel při startu applikace. Tady to právě zachytí podmínka která pak vyhodí `auth.E002`

`django/contrib/auth/management/commands/createsuperuser.py` --> Nedal jsem tomu moc času, ale řekl bych že ten check bude hlavně kvůli tomu že by se přepisovali data v `user_data`

ukázka: 
řídek 126: ```
for field_name in self.UserModel.REQUIRED_FIELDS:
            field = self.UserModel._meta.get_field(field_name)
            user_data[field_name] = options[field_name]
```

řádek 180 a 118: `user_data[self.UserModel.USERNAME_FIELD] = username`

Reply:
> Jeste jedna obecna k praci s gitem.
> 
> Kdyz jsem PR otevrel, prvni ceho jsem si vsiml bylo tohle:
> 
> ![Snímek obrazovky 2021-02-17 v 8 06 47](https://user-images.githubusercontent.com/64869/108170535-904c8180-70fa-11eb-955d-2d68c5456b69.png)
> 
> Jak to ze se v tvem PR objevuje diff na vec, ktera je uz davno v masteru? Otevrel jsem si proto SourceTree at to vidim v obrazku:
> 
> ![Snímek obrazovky 2021-02-17 v 8 06 34](https://user-images.githubusercontent.com/64869/108170641-b114d700-70fa-11eb-839e-b58681003149.png)
> 
> Tvoje vetev `david/custom_user_model` je zluta. Vsimni si ze zacina nekde dole od PR#21. `master` je ale modry, jeho aktualni stav je ta prvni modra tecka nahore. Kdyz publikujes PR ke schvaleni (tim tuplem pokud je tvuj PR vystaven delsi dobu a `master` se mezitim meni), musis pred publikovanim udelat rebase. Aby to vypadalo nejak takto:
> 
> ![Snímek obrazovky 2021-02-17 v 8 07 03](https://user-images.githubusercontent.com/64869/108170924-11a41400-70fb-11eb-9256-9394a247a194.png)
> 
> Tj. cela tva prace navazuje na posledni commit v `master` vetvi. Pak se zmeni i pohled v GitHubu, najednou tvuj reviewer dozvi, ktere soubory si opravdu zmenil ty (a ktere souvisi s tematem, ktere PR resi):
> 
> ![Snímek obrazovky 2021-02-17 v 8 08 05](https://user-images.githubusercontent.com/64869/108171068-44e6a300-70fb-11eb-8d8b-5b4d954efdf4.png)
> 
> V reci prikazu to znamena toto:
> 
> ```shell
> # vychozi stav: u sebe na lokale mam naklonovane repo
> # nemam rozvrtany zadny soubor (vsecky zmeny jsou commitnute, at uz jsi kde jsi)
> 
> # prepnu se to master vetve
> $ git checkout master
> 
> # stahnu vsecko nove z remote repa k sobe, vetve ktere u mne vyhniky at zmizi
> $ git fetch origin -p
> 
> # stahnu si up-to-date stav master vetve
> $ git pull origin master
> 
> # prepnu se zpet do sve feature vetve
> $ git checkout david/custom_user_model
> 
> # a ted to nejdulezitejsi: rebasnu se vuci masteru
> # tohle je ten krok, ktery vezme tve upravy a znovu je postupne jednu za druhou
> # "prehraje" vuci aktualni verzi masteru (ktery jsi pred chvilickou zaktualizoval na
> # posledni verzi)
> $ git rebase master
> 
> # rebase je potencionalne nebezpecny prikaz, muzes si tim kod rozvrtat
> # pokud si git neporadi sam, musis mu pomoci rebase uskutecnit (o tom nekdy priste)
> 
> # pokud je vse ok, zbyva publikovani vetve na github
> # protoze rebase ale meni historii, musis si push tentokrat vynutit --force switchem
> $ git push --force origin david/custom_user_model
> ```
> 
> A posledni vec -- protoze jsme (tedy ja jsem, pred chvili) prepsali historii vetve `david/custom_user_model`, dostal jsem te do trablu. Pokud by si ted udelal obycejny pull, asi ti na pocitaci vznikne dalsi prasopes anebo to vubec neklapne. Pull by si mel udelat budto s force switchem:
> 
> ```shell
> $ git pull --force origin  david/custom_user_model
> ```
> 
> ...anebo si svou vetev lokalne smaz a znovu checkoutni podle stavu z remote repa:
> 
> ```shell
> $ git checkout master
> $ git branch -D david/custom_user_model
> $ git checkout david/custom_user_model
> ```
> 
> Pro tip: kdyz jsem na tenkem lede, zkusenost me naucila, ze je dobre zkopirovat slozku s kompletnim projektem nekde bokem. git je svina, parkrat uz me kousla do stehna a pak jsem travil neskutecne hodiny googlenim a hledanim git-fu prikazu abych vyzehlil co jsem dobryndal. Mit po ruce stav pred tim nez se vydam na dobrodruznou vypravu je fajn.

Měl jsem za to že jsem to rebasnul. Budu si na to dávát víc pozor. Díky za hezké vysvětlení :)

Reply:
Super, diky za vysvetleni.